lib/HeroBonus.cpp

```diff
@@ -135,18 +144,36 @@ CBonusProxy & CBonusProxy::operator=(CBonusProxy && other)
 	std::swap(bonusListCachedLast, other.bonusListCachedLast);
 	std::swap(selector, other.selector);
 	std::swap(bonusList, other.bonusList);
+	std::swap(currentBonusListIndex, other.currentBonusListIndex);
+
 	return *this;
 }
 
+void CBonusProxy::swapBonusList(TConstBonusListPtr other) const
+{
+	// The idea here is to avoid changing active bonusList while it can be read by a different thread.
+	// Because such use of shared ptr is not thread safe
+	// So to avoid this we change the second offline instance and swap active index
+	auto newCurrent = 1 - currentBonusListIndex;
+	bonusList[newCurrent] = other;
+	currentBonusListIndex = newCurrent;
+}
+
 TConstBonusListPtr CBonusProxy::getBonusList() const
 {
-	if(target->getTreeVersion() != bonusListCachedLast || !bonusList)
+	if(target->getTreeVersion() != bonusListCachedLast || !bonusList[currentBonusListIndex])
 	{
+		boost::lock_guard<boost::mutex>lock(swapGuard);
+
 		//TODO: support limiters
-		bonusList = target->getAllBonuses(selector, Selector::all);
-		bonusListCachedLast = target->getTreeVersion();
+		if(target->getTreeVersion() != bonusListCachedLast || !bonusList[currentBonusListIndex])

```

Comment:
It's exactly same condition as in 164 line.

Reply:
The best practice for locking in such cases. First if is unlocked and will most likely cause most of the invokations to go else line. And if it turns out that we do need to update bonusList we enter if body, lock and verify the same condition again but under the lock already. And it might get false because another thread already updated bonusList.